File: examples/primitives/bytecode/disassembly.zig
//! Disassembly Example
//!
//! Demonstrates:
//! - Formatting bytecode as human-readable instructions
//! - Formatting individual instructions
//! - Creating disassembly output
//! - Analyzing bytecode with formatted output

const std = @import("std");
const primitives = @import("primitives");
const Bytecode = primitives.bytecode.Bytecode;
const Instruction = primitives.bytecode.Instruction;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const stdout = std.io.getStdOut().writer();

    try stdout.writeAll("\n=== Disassembly Example ===\n\n");

    // ============================================================
    // Basic Disassembly
    // ============================================================

    try stdout.writeAll("--- Basic Disassembly ---\n\n");

    const simple_hex = "0x60016002015b00";
    const simple_code = try Bytecode.fromHex(allocator, simple_hex);
    defer allocator.free(simple_code);

    try stdout.print("Bytecode: {s}\n\n", .{simple_hex});

    const disassembly = try Bytecode.formatInstructions(allocator, simple_code);
    defer {
        for (disassembly) |line| allocator.free(line);
        allocator.free(disassembly);
    }

    try stdout.writeAll("Disassembly:\n");
    for (disassembly) |line| {
        try stdout.print("  {s}\n", .{line});
    }
    try stdout.writeAll("\n");

    // ============================================================
    // Complex Bytecode Disassembly
    // ============================================================

    try stdout.writeAll("--- Complex Bytecode Disassembly ---\n\n");

    const constructor_hex = "0x608060405234801561001057600080fd5b50";
    const constructor_code = try Bytecode.fromHex(allocator, constructor_hex);
    defer allocator.free(constructor_code);

    try stdout.print("Bytecode: {s}\n\n", .{constructor_hex});

    const constructor_disasm = try Bytecode.formatInstructions(allocator, constructor_code);
    defer {
        for (constructor_disasm) |line| allocator.free(line);
        allocator.free(constructor_disasm);
    }

    try stdout.writeAll("Disassembly:\n");
    for (constructor_disasm) |line| {
        try stdout.print("  {s}\n", .{line});
    }
    try stdout.writeAll("\n");

    // ============================================================
    // Formatting Individual Instructions
    // ============================================================

    try stdout.writeAll("--- Formatting Individual Instructions ---\n\n");

    const instructions = try Bytecode.parseInstructions(allocator, simple_code);
    defer {
        for (instructions) |inst| {
            if (inst.pushData) |data| allocator.free(data);
        }
        allocator.free(instructions);
    }

    try stdout.writeAll("Individual instruction formatting:\n");
    for (instructions) |inst| {
        const formatted = try Bytecode.formatInstruction(allocator, inst);
        defer allocator.free(formatted);
        try stdout.print("  {s}\n", .{formatted});
    }
    try stdout.writeAll("\n");

    // ============================================================
    // Annotated Disassembly
    // ============================================================

    try stdout.writeAll("--- Annotated Disassembly ---\n\n");

    const annotated_hex = "0x600560565b60016002015b00";
    const annotated_code = try Bytecode.fromHex(allocator, annotated_hex);
    defer allocator.free(annotated_code);

    try stdout.print("Bytecode: {s}\n\n", .{annotated_hex});

    const analysis = try Bytecode.analyze(allocator, annotated_code);
    defer {
        for (analysis.instructions) |inst| {
            if (inst.pushData) |data| allocator.free(data);
        }
        allocator.free(analysis.instructions);
        allocator.free(analysis.jumpDestinations);
    }

    try stdout.writeAll("Annotated disassembly:\n");
    for (analysis.instructions) |inst| {
        const formatted = try Bytecode.formatInstruction(allocator, inst);
        defer allocator.free(formatted);

        try stdout.print("  {s}", .{formatted});

        // Add annotations
        var is_jumpdest = false;
        for (analysis.jumpDestinations) |pos| {
            if (pos == inst.position) {
                is_jumpdest = true;
                break;
            }
        }

        const is_terminator = Bytecode.isTerminator(inst.opcode);

        if (is_jumpdest and is_terminator) {
            try stdout.writeAll(" ; JUMP TARGET, TERMINATOR");
        } else if (is_jumpdest) {
            try stdout.writeAll(" ; JUMP TARGET");
        } else if (is_terminator) {
            try stdout.writeAll(" ; TERMINATOR");
        }

        try stdout.writeAll("\n");
    }
    try stdout.writeAll("\n");

    // ============================================================
    // Side-by-Side Comparison
    // ============================================================

    try stdout.writeAll("--- Side-by-Side Comparison ---\n\n");

    try stdout.writeAll("Pos  | Hex        | Disassembly\n");
    try stdout.writeAll("-----|------------|---------------------------\n");

    const compare_insts = try Bytecode.parseInstructions(allocator, simple_code);
    defer {
        for (compare_insts) |inst| {
            if (inst.pushData) |data| allocator.free(data);
        }
        allocator.free(compare_insts);
    }

    for (compare_insts) |inst| {
        var hex_buf: [64]u8 = undefined;
        var hex_str: []const u8 = undefined;

        if (inst.pushData) |data| {
            var fbs = std.io.fixedBufferStream(&hex_buf);
            try std.fmt.format(fbs.writer(), "0x{x:0>2} ", .{inst.opcode});
            for (data) |byte| {
                try std.fmt.format(fbs.writer(), "{x:0>2}", .{byte});
            }
            hex_str = fbs.getWritten();
        } else {
            hex_str = try std.fmt.bufPrint(&hex_buf, "0x{x:0>2}", .{inst.opcode});
        }

        const formatted = try Bytecode.formatInstruction(allocator, inst);
        defer allocator.free(formatted);

        // Extract just disassembly part (after ": ")
        var disasm: []const u8 = formatted;
        if (std.mem.indexOf(u8, formatted, ": ")) |idx| {
            disasm = formatted[idx + 2 ..];
        }

        try stdout.print("{d:4} | {s:<10} | {s}\n", .{ inst.position, hex_str, disasm });
    }
    try stdout.writeAll("\n");

    // ============================================================
    // Disassembly with PUSH Values
    // ============================================================

    try stdout.writeAll("--- Disassembly with PUSH Values ---\n\n");

    const push_hex = "0x60ff61123463abcdef01627fffffff5b00";
    const push_code = try Bytecode.fromHex(allocator, push_hex);
    defer allocator.free(push_code);

    try stdout.print("Bytecode: {s}\n\n", .{push_hex});

    const push_insts = try Bytecode.parseInstructions(allocator, push_code);
    defer {
        for (push_insts) |inst| {
            if (inst.pushData) |data| allocator.free(data);
        }
        allocator.free(push_insts);
    }

    for (push_insts) |inst| {
        const formatted = try Bytecode.formatInstruction(allocator, inst);
        defer allocator.free(formatted);

        try stdout.print("  {s}\n", .{formatted});

        if (inst.pushData) |data| {
            // Convert to value
            var value: u256 = 0;
            for (data) |byte| {
                value = (value << 8) | byte;
            }

            try stdout.print("    -> Decimal: {d}\n", .{value});
            try stdout.print("    -> Hex: 0x{x}\n", .{value});
        }
    }
    try stdout.writeAll("\n");

    // ============================================================
    // Disassembly with Jump Analysis
    // ============================================================

    try stdout.writeAll("--- Disassembly with Jump Analysis ---\n\n");

    const jump_hex = "0x6005565b60016002015b600a575b00";
    const jump_code = try Bytecode.fromHex(allocator, jump_hex);
    defer allocator.free(jump_code);

    try stdout.print("Bytecode: {s}\n\n", .{jump_hex});

    const jump_analysis = try Bytecode.analyze(allocator, jump_code);
    defer {
        for (jump_analysis.instructions) |inst| {
            if (inst.pushData) |data| allocator.free(data);
        }
        allocator.free(jump_analysis.instructions);
        allocator.free(jump_analysis.jumpDestinations);
    }

    try stdout.writeAll("Valid JUMPDEST positions: ");
    for (jump_analysis.jumpDestinations, 0..) |pos, i| {
        if (i > 0) try stdout.writeAll(", ");
        try stdout.print("{d}", .{pos});
    }
    try stdout.writeAll("\n\n");

    try stdout.writeAll("Disassembly:\n");
    for (jump_analysis.instructions) |inst| {
        var is_jumpdest = false;
        for (jump_analysis.jumpDestinations) |pos| {
            if (pos == inst.position) {
                is_jumpdest = true;
                break;
            }
        }

        const prefix = if (is_jumpdest) "→ " else "  ";

        const formatted = try Bytecode.formatInstruction(allocator, inst);
        defer allocator.free(formatted);

        try stdout.print("{s}{s}\n", .{ prefix, formatted });

        if (inst.opcode == 0x56 or inst.opcode == 0x57) {
            try stdout.writeAll("    (Jump instruction)\n");
        }
    }
    try stdout.writeAll("\n");

    // ============================================================
    // Compact Disassembly
    // ============================================================

    try stdout.writeAll("--- Compact Disassembly ---\n\n");

    const compact_hex = "0x60016002015b00";
    const compact_code = try Bytecode.fromHex(allocator, compact_hex);
    defer allocator.free(compact_code);

    const compact_insts = try Bytecode.parseInstructions(allocator, compact_code);
    defer {
        for (compact_insts) |inst| {
            if (inst.pushData) |data| allocator.free(data);
        }
        allocator.free(compact_insts);
    }

    try stdout.print("Bytecode: {s}\n", .{compact_hex});
    try stdout.writeAll("Compact: ");

    for (compact_insts, 0..) |inst, i| {
        if (i > 0) try stdout.writeAll(" ; ");

        const formatted = try Bytecode.formatInstruction(allocator, inst);
        defer allocator.free(formatted);

        // Extract just opcode part (after ": ")
        var part: []const u8 = formatted;
        if (std.mem.indexOf(u8, formatted, ": ")) |idx| {
            part = formatted[idx + 2 ..];
        }

        try stdout.writeAll(part);
    }
    try stdout.writeAll("\n\n");

    // ============================================================
    // Export Disassembly
    // ============================================================

    try stdout.writeAll("--- Export Disassembly ---\n\n");

    const export_hex = "0x60016002015b00";
    const export_code = try Bytecode.fromHex(allocator, export_hex);
    defer allocator.free(export_code);

    try stdout.writeAll("Exported disassembly:\n");
    try stdout.writeAll("; Bytecode disassembly\n");
    try stdout.print(";   Size: {d} bytes\n", .{export_code.len});
    try stdout.print(";   Hex: {s}\n", .{export_hex});
    try stdout.writeAll("\n");

    const export_disasm = try Bytecode.formatInstructions(allocator, export_code);
    defer {
        for (export_disasm) |line| allocator.free(line);
        allocator.free(export_disasm);
    }

    for (export_disasm) |line| {
        try stdout.print("{s}\n", .{line});
    }
    try stdout.writeAll("\n");

    try stdout.writeAll("=== Example Complete ===\n\n");
}

File: examples/primitives/bytecode/disassembly.ts
/**
 * Disassembly Example
 *
 * Demonstrates:
 * - Formatting bytecode as human-readable instructions
 * - Formatting individual instructions
 * - Creating disassembly output
 * - Analyzing bytecode with formatted output
 */

import { Bytecode } from "../../../src/primitives/Bytecode/index.js";

const simpleCode = Bytecode.fromHex("0x60016002015b00");

const disassembly = Bytecode.formatInstructions(simpleCode);
disassembly.forEach((line) => {});

// Solidity constructor prefix pattern
const constructorCode = Bytecode.fromHex(
	"0x608060405234801561001057600080fd5b50",
);

const constructorDisasm = Bytecode.formatInstructions(constructorCode);
constructorDisasm.forEach((line) => {});

const instructions = Bytecode.parseInstructions(simpleCode);
instructions.forEach((inst) => {
	const formatted = Bytecode.formatInstruction(inst);
});

function annotatedDisassembly(code: typeof Bytecode.prototype): string[] {
	const analysis = Bytecode.analyze(code);
	const lines: string[] = [];

	analysis.instructions.forEach((inst) => {
		let line = Bytecode.formatInstruction(inst);

		// Add annotations
		const annotations: string[] = [];

		if (analysis.jumpDestinations.has(inst.position)) {
			annotations.push("JUMP TARGET");
		}

		if (Bytecode.isTerminator(inst.opcode)) {
			annotations.push("TERMINATOR");
		}

		if (annotations.length > 0) {
			line += ` ; ${annotations.join(", ")}`;
		}

		lines.push(line);
	});

	return lines;
}

const annotatedCode = Bytecode.fromHex("0x600560565b60016002015b00");

const annotated = annotatedDisassembly(annotatedCode);
annotated.forEach((line) => {});

function sideBySide(code: typeof Bytecode.prototype): void {
	const instructions = Bytecode.parseInstructions(code);

	instructions.forEach((inst) => {
		const pos = inst.position.toString().padStart(4, " ");
		const opcodeHex = `0x${inst.opcode.toString(16).padStart(2, "0")}`;

		let hex = opcodeHex;
		if (inst.pushData) {
			const data = Array.from(inst.pushData)
				.map((b) => b.toString(16).padStart(2, "0"))
				.join("");
			hex += ` ${data}`;
		}

		const hexPadded = hex.padEnd(10, " ");
		const disasm = Bytecode.formatInstruction(inst).split(": ")[1] || "";
	});
}
sideBySide(simpleCode);

const pushCode = Bytecode.fromHex("0x60ff61123463abcdef01627fffffff5b00");

function disassemblyWithValues(code: typeof Bytecode.prototype): void {
	const instructions = Bytecode.parseInstructions(code);

	instructions.forEach((inst) => {
		const formatted = Bytecode.formatInstruction(inst);

		if (inst.pushData) {
			let value = 0n;
			for (const byte of inst.pushData) {
				value = (value << 8n) | BigInt(byte);
			}
		}
	});
}

disassemblyWithValues(pushCode);

const jumpCode = Bytecode.fromHex("0x6005565b60016002015b600a575b00");

function disassemblyWithJumps(code: typeof Bytecode.prototype): void {
	const analysis = Bytecode.analyze(code);
	const jumpdests = analysis.jumpDestinations;
	analysis.instructions.forEach((inst) => {
		let prefix = "  ";

		// Mark jump destinations
		if (jumpdests.has(inst.position)) {
			prefix = "→ ";
		}

		const formatted = Bytecode.formatInstruction(inst);

		// Annotate JUMP/JUMPI with target validation
		if (inst.opcode === 0x56 || inst.opcode === 0x57) {
		}
	});
}

disassemblyWithJumps(jumpCode);

function groupedDisassembly(code: typeof Bytecode.prototype): void {
	const analysis = Bytecode.analyze(code);
	let currentBlock = 0;

	analysis.instructions.forEach((inst, i) => {
		// Start new block at JUMPDEST
		if (analysis.jumpDestinations.has(inst.position) && i > 0) {
			currentBlock++;
		}

		const formatted = Bytecode.formatInstruction(inst);

		// End block at terminator or jump
		if (
			Bytecode.isTerminator(inst.opcode) ||
			inst.opcode === 0x56 ||
			inst.opcode === 0x57
		) {
			if (i < analysis.instructions.length - 1) {
			}
		}
	});
}

const blockedCode = Bytecode.fromHex("0x600160020100600360045b60ff5b00");

groupedDisassembly(blockedCode);

function compactDisassembly(code: typeof Bytecode.prototype): string {
	const instructions = Bytecode.parseInstructions(code);

	return instructions
		.map((inst) => {
			const formatted = Bytecode.formatInstruction(inst);
			// Extract just the opcode and data part
			const parts = formatted.split(": ");
			return parts[1] || formatted;
		})
		.join(" ; ");
}

const compactCode = Bytecode.fromHex("0x60016002015b00");
const compact = compactDisassembly(compactCode);

function disassemblyStats(code: typeof Bytecode.prototype): void {
	const analysis = Bytecode.analyze(code);
	const disasm = Bytecode.formatInstructions(code);
	disasm.forEach((line) => {});
}

const statsCode = Bytecode.fromHex("0x608060405234801561001057600080fd5b");
disassemblyStats(statsCode);

function exportDisassembly(code: typeof Bytecode.prototype): string {
	const lines: string[] = [];

	lines.push("; Bytecode disassembly");
	lines.push(`;   Size: ${Bytecode.size(code)} bytes`);
	lines.push(`;   Hex: ${Bytecode.toHex(code)}`);
	lines.push("");

	const disasm = Bytecode.formatInstructions(code);
	lines.push(...disasm);

	return lines.join("\n");
}

const exportCode = Bytecode.fromHex("0x60016002015b00");
const exported = exportDisassembly(exportCode);
